### PR TITLE
Warm the query planner cache with hot queries

### PR DESCRIPTION
**Pull request by garypen**
If we use an LRU cache (crate: lru) for our query plans, then, whenever
we update queries or configuration, we can pick the top current query
keys and use those to pre-populate a new cache.

Most of this change is just expressing that simple idea.

There are yet questions to be resolved:
 - How large should the cache be and should it be configurable? 
 - How many hot keys should we preserve on re-configuration?
 - How best to expose this functionality without affecting good
   code structure too much?

Answers:
 - Cache defaults to 100 items and is configurable
 - Preserve 20% of hot keys on re-configuration/schema update
 - Currently passed as parameter, will change to adopt configuration updates

~~Hence this is a draft PR for early review.~~
Now in final review

### CODE REVIEW

File: crates/apollo-router-core/src/query_planner/caching_query_planner.rs
Comment (Geal) [inline crates/apollo-router-core/src/query_planner/caching_query_planner.rs:48]: If the cache is large, we might block all new requests for a long time during the update. Not sure how that can be fixed here, we had RwLocks in various parts of the code before, we can reintroduce one here since writes wot be frequent

File: crates/apollo-router-core/src/traits.rs
Comment (Geal) [inline crates/apollo-router-core/src/traits.rs:11]: Since we have a type for the key, we could go the extra mile and make two structs, QueryKey and BorrowedQueryKey, to avoid cloning the suery when looking up the cache(IIRC @abernix  said some queries could be very large).

The idea is that the first struct would hold a String, Option<String> and QueryPlanOptions, and the second one &str, Option<&str> and &QueryPlanOptions. It would also implement Borrow<QueryKey> so it can be used in https://docs.rs/lru/0.7.0/lru/struct.LruCache.html#method.get (there's a similar design in HashMap).

File: crates/apollo-router/src/graph_factory.rs
Comment (Geal) [inline crates/apollo-router/src/graph_factory.rs:41]: Why remove the spawn_blocking here? It is necessary, the cal to the query planner is blocking right now, it creates and calls into a v8 vm

**Review by Geal — DISMISSED** [review]

File: crates/apollo-router-core/src/query_planner/caching_query_planner.rs
Comment (garypen) [reply crates/apollo-router-core/src/query_planner/caching_query_planner.rs:119]: The work I've done here is to avoid this problem by only waiting on keys which are currently being processed and notifying waiters when the work is complete.

File: crates/apollo-router-core/src/traits.rs
Comment (garypen) [reply crates/apollo-router-core/src/traits.rs:11]: In discussion, we agreed to leave this to a follow-up PR. I'll file an issue for this (and other outstanding work) when I finish up this PR.

File: crates/apollo-router/src/lib.rs
Comment (cecton) [inline crates/apollo-router/src/lib.rs:427]:
```suggestion
    /// Limit query cache entries.
```

File: crates/apollo-router/src/lib.rs
Comment (cecton) [inline crates/apollo-router/src/lib.rs:429]: 1.  This should go in `Configuration`

2.  I'm not convinced this is the right name for this field...
    
    This variable in particular concerns the cache size of the query planning. There will probably be more caches. For example the queries will be parsed to `Query` objects Soon :tm: and it's also kind of a query cache.

    I think we need to be more specific. For example we could create a struct `Cache` (in module `configuration`) that will have a field for each different cache. What do you think?

File: crates/apollo-router/src/router_factory.rs
Comment (cecton) [inline crates/apollo-router/src/router_factory.rs:31]: Clearly moving `query_cache_limit` to `Configuration` will reduce the number of arguments here.

File: crates/apollo-router/src/main.rs
Comment (cecton) [inline crates/apollo-router/src/main.rs:35]:
```suggestion
    /// Query cache size (number of entries).
```

**Review by cecton — CHANGES_REQUESTED** [review]
Partially reviewed. I still need to review & understand the entire logic.

File: crates/apollo-router-core/src/query_planner/caching_query_planner.rs
Comment (Geal) [reply crates/apollo-router-core/src/query_planner/caching_query_planner.rs]: erf, that was a comment in a review that was left pending, disregard that

File: crates/apollo-router-core/src/query_planner/caching_query_planner.rs
Comment (Geal) [inline crates/apollo-router-core/src/query_planner/caching_query_planner.rs:74]: repeating what I said in the outdated comment:
It might be useful to note here that in extreme cases, between the moment we receive the notification and the moment we check the cache, the value could have been evicted from the cache
so we could instead get the value from the receiver directly and avoid that race ?

File: crates/apollo-router-core/src/query_planner/caching_query_planner.rs
Comment (Geal) [inline crates/apollo-router-core/src/query_planner/caching_query_planner.rs:105]: `let results = locked_cache.keys().take(self.query_cache_limit / 5).cloned().collect::<Vec<_>>()` maybe?

Comment (cecton) [reply crates/apollo-router-core/src/query_planner/caching_query_planner.rs:105]: I already commented on that plus my solution work.... 

wait... where is my comment :thinking:

File: crates/apollo-router-core/src/query_planner/caching_query_planner.rs
Comment (cecton) [inline crates/apollo-router-core/src/query_planner/caching_query_planner.rs:108]: For some reason my comment disappeared... so here is what I said:

1. There might be unnecessary reallocation: this is because you didn't use `Vec::with_capacity`
2. This can be improved in readability using FP:

```suggestion
        locked_cache
            .iter()
            .take(self.query_cache_limit / 5)
            .map(|(key, _value)| key.clone()
            .collect()
```

File: crates/apollo-router-core/src/query_planner/caching_query_planner.rs
Comment (cecton) [reply crates/apollo-router-core/src/query_planner/caching_query_planner.rs:105]: Somehow it disappeared I don't know :woman_shrugging: Here, I recreated it: https://github.com/apollographql/router/pull/174#discussion_r756356657

(And no you can't use `.keys()` because it's not a hashmap and you shouldn't use `.cloned()` without `.keys()` because it would clone the values too. And you don't need to pass the type argument because it's inferred from the function signature)

File: crates/apollo-router-core/src/query_planner/caching_query_planner.rs
Comment (garypen) [reply crates/apollo-router-core/src/query_planner/caching_query_planner.rs:63]: Yes.

File: crates/apollo-router-core/src/query_planner/caching_query_planner.rs
Comment (garypen) [reply crates/apollo-router-core/src/query_planner/caching_query_planner.rs:74]: Great idea. I've done this.

File: crates/apollo-router-core/src/query_planner/caching_query_planner.rs
Comment (garypen) [reply crates/apollo-router-core/src/query_planner/caching_query_planner.rs:78]: I would do this, but I adopted @Geal suggestion to just return the value in the broadcast and this makes this code section redundant.

File: crates/apollo-router-core/src/query_planner/caching_query_planner.rs
Comment (Geal) [reply crates/apollo-router-core/src/query_planner/caching_query_planner.rs:105]: right, and it would not make sense to iterate on keys, since we want them in most recently used order. I guess `let results = locked_cache.iter()().take(self.query_cache_limit / 5).map(|(_k, v)| v.clone()).collect()` wuld work then

Comment (Geal) [reply crates/apollo-router-core/src/query_planner/caching_query_planner.rs:105]: huh and _now_ I see your other comment :sweat_smile:

File: crates/apollo-router-core/src/query_planner/caching_query_planner.rs
Comment (garypen) [reply crates/apollo-router-core/src/query_planner/caching_query_planner.rs:87]: You are right. I've adopted the sense of the change by not cloning when I create the key and that means I still need to clone at 87. That can lead to a reduction in cloning for most paths.

File: crates/apollo-router/src/lib.rs
Comment (garypen) [reply crates/apollo-router/src/lib.rs:429]: I know what you mean about the name and the fact that it should be in configuration (which didn't exist before my merge).
Name suggestion: plan_cache_limit (to identify that it is only for plans). I could make that change in this PR.

The bigger suggestion of changing to Configuration and the Cache struct should be in a separate Issue/PR I think, since it is probably expanding the scope of this one too much.

File: crates/apollo-router/src/main.rs
Comment (garypen) [reply crates/apollo-router/src/main.rs:35]: Done

File: crates/apollo-router-core/src/query_planner/caching_query_planner.rs
Comment (garypen) [reply crates/apollo-router-core/src/query_planner/caching_query_planner.rs:108]: Done

File: crates/apollo-router-core/src/query_planner/caching_query_planner.rs
Comment (garypen) [reply crates/apollo-router-core/src/query_planner/caching_query_planner.rs:105]: resolved in other comment.

File: crates/apollo-router/src/lib.rs
Comment (cecton) [reply crates/apollo-router/src/lib.rs]: Configuration exists for a very long time now :thinking: even before I arrived.

> Name suggestion: plan_cache_limit (to identify that it is only for plans).

Sounds good to me

> The bigger suggestion of changing to Configuration and the Cache struct should be in a separate Issue/PR I think, since it is probably expanding the scope of this one too much.

I don't think we are in a hurry and since you are introducing those I would prefer that it is included in this PR. Things that are out-of-scope imo should be things that exist already in the code and should be improved in some way. But the parameter plan_cache_limit is something introduced in this PR so it makes sense to me that this PR locates it as best it could. I'm pretty sure this will also simplify your code.

It is possible there is a reason I'm not seeing here that makes this change complicated somehow so feel free to merge if that's the case.

Comment (garypen) [reply crates/apollo-router/src/lib.rs]: Sorry, that was my mis-apprehension. I was trying to refer to the changes you had made to internal parameters yesterday.

I could add it as a configuration, but I slightly prefer a command line and I think it could be resolved in a later issue/PR.

I think there's a good discussion here which will keep to a later issue/PR, so I will merge this and preserve the discussion in issue #188

Comment (garypen) [reply crates/apollo-router/src/lib.rs]: Resolving as moved to #188

File: crates/apollo-router/src/router_factory.rs
Comment (garypen) [reply crates/apollo-router/src/router_factory.rs]: Resolving as part of #188

File: crates/apollo-router-core/src/query_planner/caching_query_planner.rs
Comment (cecton) [inline crates/apollo-router-core/src/query_planner/caching_query_planner.rs:109]: This needs to move outside the trait implementation.

The trait QueryPlanner shouldn't be aware of any caching functionality.

```suggestion
}

impl<T: QueryPlanner> CachingQueryPlanner<T> { 
    async fn get_hot_keys(&self) -> Vec<QueryKey> {
        let locked_cache = self.cached.lock().await;
        locked_cache
            .iter()
            .take(self.plan_cache_limit / 5)
            .map(|(key, _value)| key.clone())
            .collect()
    }
}
```

**Review by cecton — CHANGES_REQUESTED** [review]
So as the discussion about the parameter as been moved, this is not blocking this PR anymore :grin: 

I still have not finished reviewing the core logic though but I have more remarks about the traits.

**Review by o0Ignition0o — APPROVED** [review]
The only question I had was about the get_hot_keys function, which we cleared up together, otherwise this looks good to me overall.

The conversations seem to be more about nits, which can be iterated on IMO

File: crates/apollo-router-core/src/query_planner/caching_query_planner.rs
Comment (garypen) [reply crates/apollo-router-core/src/query_planner/caching_query_planner.rs:109]: I agree that it may be undesirable, but I think this would break RouterFactory::recreate() because: 
```let hot_keys = graph.get_query_planner().get_hot_keys().await;```
I'm trying to decide if there's a reason why a non-caching QueryPlanner would return hot keys... I mean, it could, but the current implementation doesn't do any tracking and just returns an empty vector.

File: crates/apollo-router-core/src/query_planner/caching_query_planner.rs
Comment (cecton) [inline crates/apollo-router-core/src/query_planner/caching_query_planner.rs:74]: According to the [documentation](https://docs.rs/tokio/1.14.0/tokio/sync/broadcast/error/enum.RecvError.html) there are only 2 possible errors:

1. The channel was closed: "There are no more active senders implying no further messages will ever be sent."
2. "The receiver lagged too far behind. Attempting to receive again will return the oldest message still retained by the channel."

I guess point 2. makes no sense because only one message will ever be sent to this channel.

Point 1 happens only if all the senders are dropped but there is a receiver awaiting somewhere. This won't happen because the locked_wait_map will be locked while the message is being sent to the channel so no other receiver can be created during that time:

```rust
                // ...
                let mut locked_wait_map = self.wait_map.lock().await;
                locked_wait_map.remove(&key);
                // Let our waiters know
                let broadcast_value = value.clone();
                tokio::task::spawn_blocking(move || tx.send((key, broadcast_value))).await??;
                value
            } // locked_wait_map is dropped here
```

So I think you can safely do `.expect()` on this PR and explain the proof.

We don't want to add the RecvError to the QueryPlannerError as this can be returned to the user making the call. So either we make a strategy to "retry" to prevent failing or we unwrap because we are 100% confident that this will not fail. Correct me if I'm wrong but I think it's pretty clear that we can prove this won't ever fail.

File: crates/apollo-router/src/router_factory.rs
Comment (garypen) [reply crates/apollo-router/src/router_factory.rs:75]: I'm calling a bunch of async functions and recreate() is not async.

(I might be missing something here...)

File: .gitignore
Comment (garypen) [reply .gitignore]: done

File: crates/apollo-router-core/src/query_planner/caching_query_planner.rs
Comment (cecton) [reply crates/apollo-router-core/src/query_planner/caching_query_planner.rs:109]: Yes I just saw... :thinking:

File: crates/apollo-router/src/router_factory.rs
Comment (cecton) [inline crates/apollo-router/src/router_factory.rs:92]: `recreate` actually is async because it returns a BoxFuture. The `async` syntax of Rust is just sugar for: `impl Future<Output = RESULT> + Send>` (more or less).


```suggestion
        Box::pin(async move {
            // Use the "hot" entries in the supplied graph to pre-populate
            // our new graph
            let new_graph = factory.await;
            let hot_keys = graph.get_query_planner().get_hot_keys().await;
            // It would be nice to get these keys concurrently by spawning
            // futures in our loop. However, these calls to get call the
            // v8 based query planner and running too many of these
            // concurrently is a bad idea. One for the future...
            for key in hot_keys {
                // We can ignore errors, since we are just warming up the
                // cache
                let _ = new_graph.get_query_planner().get(key.0, key.1, key.2).await;
            }
            new_graph
        })
```

File: crates/apollo-router-core/src/query_planner/caching_query_planner.rs
Comment (cecton) [inline crates/apollo-router-core/src/query_planner/caching_query_planner.rs:95]: According to the [documentation](https://docs.rs/tokio/1.14.0/tokio/sync/broadcast/struct.Sender.html#method.send):

> Attempts to send a value to all active Receiver handles, returning it back if it could not be sent.
> 
> A successful send occurs when there is at least one active Receiver handle. An unsuccessful send would be one where all associated Receiver handles have already been dropped.

It is possible that the receiver have been canceled for some reason. But this is no reason to fail the current request. I think you can just safely ignore this error:

```rust
let _ = tx.send((key, broadcast_value))
```

Again, just like RecvError, we don't want to add the SendError to the QueryPlannerError as this can be returned to the user making the call. So either we make a strategy to "retry" to prevent failing or we unwrap because we are 100% confident that this will not fail. Correct me if I'm wrong but I think here we can just ignore any error from sending the message.

**Review by cecton — CHANGES_REQUESTED** [review]

File: crates/apollo-router/src/router_factory.rs
Comment (garypen) [reply crates/apollo-router/src/router_factory.rs:92]: Ah, thank you! Done.

File: crates/apollo-router-core/src/query_planner/caching_query_planner.rs
Comment (garypen) [reply crates/apollo-router-core/src/query_planner/caching_query_planner.rs:79]: No good reason. I'll set it to 1.

File: crates/apollo-router/src/router_factory.rs
Comment (garypen) [reply crates/apollo-router/src/router_factory.rs:75]: resolved in later comment.

File: crates/apollo-router-core/src/query_planner/caching_query_planner.rs
Comment (garypen) [reply crates/apollo-router-core/src/query_planner/caching_query_planner.rs:76]: Done

File: crates/apollo-router-core/src/query_planner/caching_query_planner.rs
Comment (garypen) [reply crates/apollo-router-core/src/query_planner/caching_query_planner.rs:74]: I agree that the documentation suggest this is safe to do, so I've done it.

File: crates/apollo-router-core/src/query_planner/caching_query_planner.rs
Comment (garypen) [reply crates/apollo-router-core/src/query_planner/caching_query_planner.rs:95]: (Repeat from above) I agree that the documentation suggest this is safe to do, so I've done it.

File: crates/apollo-router-core/src/query_planner/caching_query_planner.rs
Comment (cecton) [inline crates/apollo-router-core/src/query_planner/caching_query_planner.rs:79]: You can actually put the proof message in the expect.

```suggestion
                let (recv_key, recv_plan) = receiver.recv().await
                    .expect("the sender won't ever be dropped before all the receivers finish; qed");
```

File: crates/apollo-router-core/src/query_planner/caching_query_planner.rs
Comment (cecton) [inline crates/apollo-router-core/src/query_planner/caching_query_planner.rs:106]:
```suggestion
                tokio::task::spawn_blocking(move || {
                    tx.send((key, broadcast_value))
                        .expect("there is always at least one receiver alive, the _rx guard; qed")
                })
```

File: crates/apollo-router/src/router_factory.rs
Comment (cecton) [inline crates/apollo-router/src/router_factory.rs:68]: This thing has been renamed to "router".

```suggestion
        router: Arc<ApolloRouter>,
```

File: crates/apollo-router/src/router_factory.rs
Comment (cecton) [inline crates/apollo-router/src/router_factory.rs:26]: This thing has been renamed  `router`.
```suggestion
        router: Arc<Router>,
```

File: crates/apollo-router-core/src/query_planner/caching_query_planner.rs
Comment (cecton) [reply crates/apollo-router-core/src/query_planner/caching_query_planner.rs:121]: I found the API problem behind this issue but it's best to solve it in another PR. So let keep this for now.

**Review by cecton — APPROVED** [review]
Some small knits but good to go :hugs: Thanks a lot for your patience!!

File: crates/apollo-router-core/src/query_planner/caching_query_planner.rs
Comment (garypen) [reply crates/apollo-router-core/src/query_planner/caching_query_planner.rs:79]: done

File: crates/apollo-router-core/src/query_planner/caching_query_planner.rs
Comment (garypen) [reply crates/apollo-router-core/src/query_planner/caching_query_planner.rs:106]: done

File: crates/apollo-router/src/router_factory.rs
Comment (garypen) [reply crates/apollo-router/src/router_factory.rs:68]: done

File: crates/apollo-router/src/router_factory.rs
Comment (garypen) [reply crates/apollo-router/src/router_factory.rs:26]: done

**Comment by garypen** [comment]
> Some small knits but good to go 🤗 Thanks a lot for your patience!!

Thanks for your diligent review. I think the quality is much improved.

**Pull request merged**
